### PR TITLE
312 accept types in internal csv reading function

### PR DESCRIPTION
**Pull request by folmos-at-orange**
This will enable an easier implementation of #74 and #74 in #306. 

```
commit 149fb30b1ed6f7e6562d1405d38a02388c6a56dc (HEAD -> 312-accept-types-in-internal-csv-reading-function, origin/312-accept-types-in-internal-csv-reading-function)
Author: Felipe Olmos <92923444+folmos-at-orange@users.noreply.github.com>
Date:   Tue Dec 17 17:01:40 2024 +0100

    Control the types on sklearn internal read table

    This is done only for KhiopsClassifier and KhiopsRegressor. It is
    critical for KhiopsClassifier as it accepts many target types.

    For KhiopsEncoder and KhiopsCoclustering is less critical and for the
    first one it is very complex.

commit 69dce771f41cf2c88f2cd8f757752e9f09f43854
Author: Felipe Olmos <92923444+folmos-at-orange@users.noreply.github.com>
Date:   Tue Dec 17 17:05:22 2024 +0100

    Fix a pylint nit

commit 9fafd2acd433330a9caf9712ee5b67e986d4928c
Author: Felipe Olmos <92923444+folmos-at-orange@users.noreply.github.com>
Date:   Tue Dec 17 17:13:16 2024 +0100

    Rename transform_pairs parameter to transform_type_pairs
```

---

### TODO Before Asking for a Review
- [x] Rebase your branch to the latest version of `dev` (or `main` for release PRs)
- [x] Make sure all CI workflows are green
- [x] When adding a public feature/fix: Update the `Unreleased` section of `CHANGELOG.md` (no date)
- [x] Self-Review: Review "Files Changed" tab and fix any problems you find
- API Docs (only if there are changes in docstrings, rst files or samples):
  - [x] Check the docs build **without** warning: see the log of the API Docs workflow
  - [x] Check that your changes render well in HTML: download the API Docs artifact and open `index.html`
  - If there are any problems it is faster to iterate by [building locally the API Docs](../blob/dev/doc/README.md#build-the-documentation)

### CODE REVIEW

File: khiops/sklearn/estimators.py [hunk 1]
```diff
@@ -478,11 +478,13 @@ def _transform(
         self._transform_check_dataset(ds)
 
         # Create a deployment dataset
-        # Note: The input dataset is not necessarily ready to be deployed
+        # Note: The input dataset isn't ready for deployment in the case of coclustering
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:481]: I would rephrase slightly:
`# Note: The input dataset isn't always ready for deployment, e.g. in the case of coclustering`

File: khiops/sklearn/estimators.py [hunk 2]
```diff
@@ -493,10 +495,36 @@ def _transform(
...
+
+            # On multi-table:
+            # - Reorder the table to the original table order
+            #     - Because transformed data table file is sorted by key
+            # - Drop the key columns if specified
+            if deployment_ds.is_multitable:
+                key_df = deployment_ds.main_table.data_source[
+                    deployment_ds.main_table.key
+                ]
+                output_table_df_or_path = key_df.merge(
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:514]: Why `_or_path` here, as `deployment_ds` is in-memory?

File: khiops/sklearn/estimators.py [hunk 3]
```diff
@@ -1366,16 +1356,12 @@ def _transform_create_deployment_dataset(self, ds, computation_dir):
...
-    def _transform_deployment_post_process(
-        self, deployment_ds, output_table_path, drop_key
-    ):
-        assert deployment_ds.is_multitable
-        return super()._transform_deployment_post_process(
-            deployment_ds, output_table_path, drop_key
-        )
+    def _transform_prepare_deployment_for_predict(self, _):
+        # TODO
+        #  Replace the second return value (the output columns' dtypes) with a proper
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:1361]: I would write as `# TODO: Replace the second return value ...` so that `TODO` occurs on the same line as the beginning of what is to be done (practical for `grep`ping `TODO`s to have a first glance of their status).

File: khiops/sklearn/estimators.py [hunk 4]
```diff
@@ -3115,7 +3117,11 @@ def _transform_prepare_deployment_model(self, ds):
         if self.model_target_variable_name_ not in list(ds.main_table.column_ids):
             model_dictionary.remove_variable(self.model_target_variable_name_)
 
-        return model_copy
+        # TODO
```
Comment (popescu-v) [inline khiops/sklearn/estimators.py:3120]: I would rewrite to put `TODO` on the same line as the start of what follows (to facilitate searches and look-ups) - see the same comment above.

File: khiops/utils/dataset.py [hunk 5]
```diff
@@ -357,18 +357,34 @@ def read_internal_data_table(file_path_or_stream):
...
     `pandas.DataFrame`
-        The dataframe representation.
+        The dataframe representation of the data table.
     """
+    # Change the 'U' types (Unicode strings) to 'O' because pandas does not support them
+    # in read_csv
+    if column_dtypes is not None:
+        execution_column_dtypes = {}
+        for column_name, dtype in column_dtypes.items():
+            if hasattr(dtype, "kind") and dtype.kind == "U":
```
Comment (popescu-v) [inline khiops/utils/dataset.py:382]: Why not write this `if` more concisely as `if getattr(dtype, "kind", None) == "U"`?

File: khiops/utils/dataset.py [hunk 6]
```diff
@@ -1132,6 +1149,9 @@ def __repr__(self):
             f"dtypes={dtypes_str}>"
         )
 
+    def get_column_dtype(self, column_id):
+        return self.data_source.dtypes[column_id]
```
Comment (popescu-v) [inline khiops/utils/dataset.py:1153]: I'm not sure I see the added value of using `.get_column_dtype(column_id)`  vs. using `.data_source.dtypes[column_id]`, as long as there is no further check / protection against using a non-existing `column_id`.

File: tests/test_sklearn_output_types.py [hunk 7]
```diff
@@ -50,21 +51,48 @@ class KhiopsSklearnOutputTypes(unittest.TestCase):
     def setUp(self):
         KhiopsTestHelper.skip_long_test(self)
 
+    def _replace(self, array, replacement_dict):
+        return np.array([replacement_dict[value] for value in array])
```
Comment (popescu-v) [inline tests/test_sklearn_output_types.py:55]: Should we check that `value` is in `replacement_dict` for any `value` in `array`?

File: khiops/sklearn/estimators.py (hunk 1 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:481]: It is more precise as it is. The coclustering is the only case when it is not ready.

File: khiops/sklearn/estimators.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:514]: `output_df_or_path` is the output variable. In this case we are in the `df` branch. I see no need to add additional variables.

File: khiops/sklearn/estimators.py (hunk 4 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:3120]: Ok & Done.

File: khiops/utils/dataset.py (hunk 5 above)
Comment (folmos-at-orange) [reply khiops/utils/dataset.py:382]: I don't see the gain of the conciseness. It is clear as it is.

File: khiops/utils/dataset.py (hunk 6 above)
Comment (folmos-at-orange) [reply khiops/utils/dataset.py:1153]: Added an assert.

File: tests/test_sklearn_output_types.py (hunk 7 above)
Comment (folmos-at-orange) [reply tests/test_sklearn_output_types.py:55]: I don't think so: This is a test, and we'll have a `KeyError` if there are any programming errors.

File: khiops/utils/dataset.py (hunk 6 above)
Comment (popescu-v) [reply khiops/utils/dataset.py]: Is an `assert` the right mechanism here, knowing that this is a "public" method?

File: tests/test_sklearn_output_types.py (hunk 7 above)
Comment (popescu-v) [reply tests/test_sklearn_output_types.py:55]: Ok, to me, the key insight here is "in a test suite".

File: khiops/utils/dataset.py (hunk 5 above)
Comment (popescu-v) [reply khiops/utils/dataset.py]: Ok. As this is a stylistic remark, I'll resolve this conversation.

File: khiops/sklearn/estimators.py (hunk 2 above)
Comment (popescu-v) [reply khiops/sklearn/estimators.py:514]: Yes, I can see this, but I would have appreciated a more relevant (local) name. But as this is a stylistic remark, I will resolve this conversation.

File: khiops/sklearn/estimators.py (hunk 1 above)
Comment (popescu-v) [reply khiops/sklearn/estimators.py:481]: Ok.

File: khiops/sklearn/estimators.py (hunk 2 above)
Comment (folmos-at-orange) [reply khiops/sklearn/estimators.py:514]: It will be gone when we eliminate the file path dataset support anyway.

File: khiops/utils/dataset.py (hunk 6 above)
Comment (folmos-at-orange) [reply khiops/utils/dataset.py]: Right. Changed to `KeyError` with custom message.

File: khiops/sklearn/dataset.py [hunk 8]
```diff
@@ -1132,6 +1149,11 @@ def __repr__(self):
             f"dtypes={dtypes_str}>"
         )
 
+    def get_column_dtype(self, column_id):
+        if column_id not in self.data_source.dtypes:
+            raise KeyError(f"Column '{column_id}' not found in dtypes field")
```
Comment (popescu-v) [inline khiops/sklearn/dataset.py:1154]: s/in dtypes field/in the dtypes field/

**Review by popescu-v — APPROVED** [review]

File: khiops/sklearn/dataset.py (hunk 8 above)
Comment (folmos-at-orange) [reply khiops/sklearn/dataset.py:1154]: Ok
